eth/utils/bls.py

```diff
@@ -18,117 +21,163 @@
     FQ12,
     pairing,
     normalize,
-    field_modulus,
+    field_modulus as q,
     b,
     b2,
     is_on_curve,
     curve_order,
     final_exponentiate
 )
-from eth.utils.blake import blake
+from eth_hash.auto import keccak as hash
+
 
+G2_cofactor = 305502333931268344200999753193121504214466019254188142667664032982267604182971884026507427359259977847832272839041616661285803823378372096355777062779109  # noqa: E501
+qmod = q ** 2 - 1
+eighth_roots_of_unity = [
+    FQ2([1, 1]) ** ((qmod * k) // 8)
+    for k in range(8)
+]
 
-CACHE = {}  # type: Dict[bytes, Tuple[FQ2, FQ2, FQ2]]
-# 16th root of unity
-HEX_ROOT = FQ2([21573744529824266246521972077326577680729363968861965890554801909984373949499,
-                16854739155576650954933913186877292401521110422362946064090026408937773542853])
 
+#
+# Helpers
+#
+def modular_squareroot(value: int) -> FQ2:
+    """
+    ``modular_squareroot(x)`` returns the value ``y`` such that ``y**2 % q == x``,
+    and None if this is not possible. In cases where there are two solutions,
+    the value with higher imaginary component is favored;
+    if both solutions have equal imaginary component the value with higher real
+    component is favored.
+    """
+    candidate_squareroot = value ** ((qmod + 8) // 16)
+    check = candidate_squareroot ** 2 / value
+    if check in eighth_roots_of_unity[::2]:
+        x1 = candidate_squareroot / eighth_roots_of_unity[eighth_roots_of_unity.index(check) // 2]
+        x2 = FQ2([-x1.coeffs[0], -x1.coeffs[1]])
+        # x2 = - x2
+        return x1 if (x1.coeffs[1], x1.coeffs[0]) > (x2.coeffs[1], x2.coeffs[0]) else x2
```

Comment:
Oh nice catch 👀
The `x1` and `x2` are `FQ2` objects, and `coeffs` is `Sequence[Union[int, "FQ"]]`. So
1. If `FQ2` is initialized with `int` -> no `FQ2_object.coeffs[1].n` element
2. If `FQ2` is initialized wtih `FQ` -> there is `FQ2_object.coeffs[1].n` element

Since it is `int` in our usage, without `.n` is correct now.

I'll note it in the spec repo.